### PR TITLE
Remove reflection-based lightup S.R.M for netcoreapp

### PR DESCRIPTION
**Pull request by jkotas**
S.R.Metadata has netcoreapp specific build. We can take advantage of it to avoid reflection and trial and error (avoids exception to be thrown and swallowed in some cases). The existing logic is preserved for the netstandard builds.

### CODE REVIEW

File: src/System.Reflection.Metadata/src/System/Reflection/Internal/Utilities/EncodingHelper.netcoreapp.cs [hunk 1]
```diff
@@ -0,0 +1,77 @@
...
+            fixed (byte* prefixedBytes = &buffer[0])
+            {
+                result = utf8Decoder.GetString(prefixedBytes, prefixedByteCount);
+            }
+
+            ArrayPool<byte>.Shared.Return(buffer);
+            return result;
+        }
+
+        public static string GetString(this Encoding encoding, byte* bytes, int byteCount)
```
Comment (tmat) [inline src/System.Reflection.Metadata/src/System/Reflection/Internal/Utilities/EncodingHelper.netcoreapp.cs]: >public static string GetString(this Encoding encoding, byte* bytes, int byteCount) [](start = 8, length = 82)

Is this method needed? If Encoding.GetString is available the call site should pick it up without needing an extension method.

File: src/System.Reflection.Metadata/src/System/Reflection/Internal/Utilities/FileStreamReadLightUp.cs [hunk 2]
```diff
@@ -142,33 +80,16 @@ internal static unsafe bool TryReadFile(Stream stream, byte* buffer, long start,
...
-                 byte* buffer,
-                 int byteCount,
-                 out int bytesRead,
-                 IntPtr overlapped
-            );
-        }
+#pragma warning disable BCL0015 // Disable Pinvoke analyzer errors.
+        [DllImport(@"kernel32.dll", EntryPoint = "ReadFile", ExactSpelling = true, SetLastError = true)]
+        [return: MarshalAs(UnmanagedType.Bool)]
+        internal static unsafe extern bool ReadFileCompat(
```
Comment (tmat) [inline src/System.Reflection.Metadata/src/System/Reflection/Internal/Utilities/FileStreamReadLightUp.cs]: >ReadFileCompat [](start = 43, length = 14)

Can rename to ReadFile as there aren't two versions anymore.

**Comment by tmat** [comment]
:shipit:

File: src/System.Reflection.Metadata/src/System/Reflection/Internal/Utilities/FileStreamReadLightUp.cs [hunk 3]
```diff
@@ -142,33 +79,16 @@ internal static unsafe bool TryReadFile(Stream stream, byte* buffer, long start,
...
-            [return: MarshalAs(UnmanagedType.Bool)]
-            internal static extern bool ReadFileCompat(
-                 SafeHandle fileHandle,
-                 byte* buffer,
-                 int byteCount,
-                 out int bytesRead,
-                 IntPtr overlapped
-            );
-        }
+#pragma warning disable BCL0015 // Disable Pinvoke analyzer errors.
```
Comment (danmosemsft) [inline src/System.Reflection.Metadata/src/System/Reflection/Internal/Utilities/FileStreamReadLightUp.cs]: I would prefer any BCL0015 be avoided and instead listed in PinvokeAnalyzerExceptionList.analyzerdata so they are in one place and not overlooked.

For ReadFile in particular it is already in both OneCore and UWP approved lists so this should not be necessary anyway.

Comment (danmosemsft) [inline src/System.Reflection.Metadata/src/System/Reflection/Internal/Utilities/FileStreamReadLightUp.cs]: The only other remaining ones are in DirectoryServices and SqlClient  and those should go inbaselines too. (I can do this)

Comment (jkotas) [inline src/System.Reflection.Metadata/src/System/Reflection/Internal/Utilities/FileStreamReadLightUp.cs]: > For ReadFile in particular it is already in both OneCore and UWP approved list

That's what I thought too. But the build still complained when I have removed it.

> I can do this

Thanks. I do not plan to do anything more for this.

**Pull request merged**